base/special/rem_pio2.jl

```diff
@@ -0,0 +1,338 @@
+# This file is a part of Julia. Except for the rem_pio2_kernel, and
+# cody_waite_* methods (see below) license is MIT: https://julialang.org/license
+
+# rem_pio2_kernel and cody_waite_* methods are heavily based on FDLIBM code:
+# __ieee754_rem_pio2, that is made available under the following licence:
+
+## Copyright (C) 1993 by Sun Microsystems, Inc. All rights reserved.
+##
+## Developed at SunPro, a Sun Microsystems, Inc. business.
+## Permission to use, copy, modify, and distribute this
+## software is freely granted, provided that this notice
+## is preserved.
+
+const invpio2 =  6.36619772367581382433e-01
+const pio2_1  =  1.57079632673412561417e+00
+const pio2_1t =  6.07710050650619224932e-11
+const pio2_2  =  6.07710050630396597660e-11
+const pio2_2t =  2.02226624879595063154e-21
+const pio2_3  =  2.02226624871116645580e-21
+const pio2_3t =  8.47842766036889956997e-32
+
+# Bits of 1/2π
+#   1/2π == sum(x / 0x1p64^i for i,x = enumerate(INV2PI))
+# Can be obtained by:
+#
+#    setprecision(BigFloat, 4096)
+#    I = 0.5/big(pi)
+#    for i = 1:19
+#        I *= 0x1p64
+#        k = trunc(UInt64, I)
+#        @printf "0x%016x,\n" k
+#        I -= k
+#    end
+const INV2PI = UInt64[
+    0x28be_60db_9391_054a,
+    0x7f09_d5f4_7d4d_3770,
+    0x36d8_a566_4f10_e410,
+    0x7f94_58ea_f7ae_f158,
+    0x6dc9_1b8e_9093_74b8,
+    0x0192_4bba_8274_6487,
+    0x3f87_7ac7_2c4a_69cf,
+    0xba20_8d7d_4bae_d121,
+    0x3a67_1c09_ad17_df90,
+    0x4e64_758e_60d4_ce7d,
+    0x2721_17e2_ef7e_4a0e,
+    0xc7fe_25ff_f781_6603,
+    0xfbcb_c462_d682_9b47,
+    0xdb4d_9fb3_c9f2_c26d,
+    0xd3d1_8fd9_a797_fa8b,
+    0x5d49_eeb1_faf9_7c5e,
+    0xcf41_ce7d_e294_a4ba,
+    0x9afe_d7ec_47e3_5742,
+    0x1580_cc11_bf1e_daea]
+
+"""
+    highword(x)
+
+Return the high word of `x` as a `UInt32`.
+"""
+@inline highword(x::UInt64) = unsafe_trunc(UInt32,x >> 32)
+@inline highword(x::Float64) = highword(reinterpret(UInt64, x))
+
+"""
+    poshighword(x)
+
+Return positive part of the high word of `x` as a `UInt32`.
+"""
+@inline poshighword(x::UInt64) = unsafe_trunc(UInt32,x >> 32)&0x7fffffff
+@inline poshighword(x::Float64) = poshighword(reinterpret(UInt64, x))
+
+"""
+    rint(x::Float64)
+
+Rounds `x` to the nearest integer, tie-breaking to even. Used internally
+in rem_pio2_kernel.
+"""
+function rint(x::Float64)
+    rf = 1.5/eps(Float64)
+    (x+rf)-rf
+end
+# This could be replaced with:
```

Comment:
I suspect that `round` might be faster when the architecture has the `ROUNDSD` instruction (introduced SSE 4.1). Before that, I think the add/subtract will probably be faster (as it will then require a libm call).